lib/record_store/provider/ns1.rb

```diff
@@ -0,0 +1,192 @@
+require 'ns1'
+
+module RecordStore
+  class Provider::NS1 < Provider
+    class << self
+      def client
+        Provider::NS1::Client.new(api_key: secrets['api_key'])
+      end
+
+      # Downloads all the records from the provider.
+      #
+      # Returns: an array of `Record` for each record in the provider's zone
+      def retrieve_current_records(zone:, stdout: $stdout)
+        records_for_zone(zone)
+         .map{|short_record| build_from_api(short_record, zone)}
+         .flatten
+      end
+
+      # Returns an array of the zones managed by provider as strings
+      def zones
+        client.zones.map do |zone|
+          zone["zone"]
+        end
+      end
+
+      private
+
+      # Fetches simplified records for the provided zone
+      def records_for_zone(zone)
+        client.zone(zone)["records"]
+      end
+
+      # Creates a new record to the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def add(record, zone)
+        new_answers = [{ answer: [ build_api_answer_from_record(record) ]}]
+
+        record_fqdn = record.fqdn.sub(/.$/, '')
+
+        existing_record = client.record(
+          zone: zone,
+          fqdn: record_fqdn,
+          type: record.type
+        )
+
+        if existing_record.nil?
+          client.create_record(
+            zone: zone,
+            fqdn: record_fqdn,
+            type: record.type,
+            params: { answers: new_answers, ttl: record.ttl }
+          )
+          return
+        end
+
+        existing_answers = existing_record["answers"].map { |answer| symbolize_keys(answer) }
+        client.modify_record(
+          zone: zone,
+          fqdn: record_fqdn,
+          type: record.type,
+          params: { answers: existing_answers + new_answers }
+        )
+      end
+
+      # Deletes an existing record from the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def remove(record, zone)
+        record_fqdn = record.fqdn.sub(/.$/, '')
+
+        existing_record = client.record(
+          zone: zone,
+          fqdn: record_fqdn,
+          type: record.type
+        )
+        return if existing_record.nil?        
+        
+        pruned_answers = existing_record["answers"]
+          .map { |answer| symbolize_keys(answer) }
+          .reject { |answer| answer[:answer] == build_api_answer_from_record(record) } 
+
+        if pruned_answers.empty?
+          client.delete_record(
+            zone: zone,
+            fqdn: record_fqdn,
+            type: record.type
+          )
+          return
+        end
+
+        client.modify_record(
+          zone: zone,
+          fqdn: record_fqdn,
+          type: record.type,
+          params: { answers: pruned_answers }
```

Comment:
If the TTL is computed, from the answers' TTLs, it would maybe change here ?

Reply:
This is an interesting point.  It's impossible to have, say, two A records for a single fqdn with different TTLs.  We'll want to be deliberate about how to handle attempts to change a TTL.  We could either raise when inconsistent TTLs are encountered (probably a smart choice) or update the TTL to that of the latest record for an FQDN and type.

Reply:
Ah, I left a comment here but I guess it's all gone without saving once I canceled review mode or something (all comments were pending). I left the comment with similar idea of Chris here. I also quickly tested with inconsistent TTLs, and the TTL used for the first record always stays.

Reply:
While working through this, we came to the following conclusions:

- If the provider's asked to `add` or `update` a record, it will have been provided a desired `RecordStore::Record` (complete with desired TTL), and will therefore be able to set the TTL of the record in NS1 to at least some desired TTL for records with that fqdn and type.
- If the provider's asked to remove a record, it will only have been provided the `RecordStore::Record` (complete with TTL) for a record that we want to discard.  There isn't enough information to know any of the other desired TTLs.  So, if someone removes one record from an answer set, the TTL will remain unchanged.  We'll have to figure out if this a huge problem or not going forward, but...it's kinda a corner case.